bundles/org.openhab.binding.tapocontrol/src/main/java/org/openhab/binding/tapocontrol/internal/TapoDiscoveryService.java

```
@@ -0,0 +1,240 @@
+    /**
+     * CREATE DISCOVERY RESULT
+     * creates discoveryResult (Thing) from JsonObject got from Cloud
+     * 
+     * @param device JsonObject with device information
+     * @return DiscoveryResult-Object
+     */
+    public DiscoveryResult createResult(JsonObject device) {
+        try {
+            String deviceModel = getDeviceModel(device);
+            String label = getDeviceLabel(device);
+            String name = device.get(CLOUD_PROPERTY_NAME).getAsString();
+            String deviceMAC = device.get(CLOUD_PROPERTY_MAC).getAsString();
+            ThingUID bridgeUID = bridge.getThing().getUID();
+            ThingTypeUID thingTypeUID = new ThingTypeUID(BINDING_ID, deviceModel);
+            ThingUID thingUID = new ThingUID(thingTypeUID, bridgeUID, deviceMAC);
+
+            /* create properties */
+            Map<String, Object> properties = new HashMap<>(1);
+            properties.put(Thing.PROPERTY_VENDOR, DEVICE_VENDOR);
+            properties.put(Thing.PROPERTY_MAC_ADDRESS, formatMac(deviceMAC, MAC_DIVISION_CHAR));
+            properties.put(Thing.PROPERTY_FIRMWARE_VERSION, device.get(CLOUD_PROPERTY_FW).getAsString());
+            properties.put(Thing.PROPERTY_HARDWARE_VERSION, device.get(CLOUD_PROPERTY_HW).getAsString());
+            properties.put(Thing.PROPERTY_MODEL_ID, deviceModel);
+            properties.put(Thing.PROPERTY_SERIAL_NUMBER, device.get(CLOUD_PROPERTY_ID).getAsString());
+
+            logger.debug("device {} discovered", deviceModel);
+            DiscoveryResult discoveryResult = DiscoveryResultBuilder.create(thingUID).withProperties(properties)
+                    .withRepresentationProperty(DEVICE_REPRASENTATION_PROPERTY).withBridge(bridgeUID).withLabel(label)
+                    .build();
+            return discoveryResult;
+        } catch (Exception e) {
+            logger.error("error creating discoveryResult", e);
+            return DiscoveryResultBuilder.create(new ThingUID(BINDING_ID, "")).build();
```

You could throw the exception. Then, you don't need to fire an empty Thing during discovery.

---

Can you comment on this?